mitiq/observable/observable.py

```diff
@@ -154,3 +179,30 @@ def _expectation_from_density_matrix(
 
     def __str__(self) -> str:
         return " + ".join(map(str, self._paulis))
+
+    def __eq__(self, other: Any) -> bool:
+        return np.allclose(self.matrix(), other.matrix())
+
+
+def _combine_duplicate_pauli_strings(
+    paulis: list[PauliString],
+) -> list[PauliString]:
+    """Combines duplicate PauliStrings by adding their coefficients.
+    Discards paulis with zero coefficients.
+
+    Returns: deduped list of PauliStrings.
+    """
+    pauli_string_coefficients: defaultdict[PauliString, complex] = defaultdict(
+        complex
+    )
+    for pauli_string in paulis:
+        cache_key = pauli_string.with_coeff(1)
+        new_coeff = pauli_string.coeff + pauli_string_coefficients[cache_key]
+        pauli_string_coefficients[cache_key] = new_coeff
```

Comment:
I think we can use the `defaultdict` functionality here to simplify the new coefficient calculation:
```suggestion
    for pauli_string in paulis:
        cache_key = pauli_string.with_coeff(1)
        pauli_string_coefficients[cache_key] += pauli_string.coeff
```